Convert exception throwing JIT helpers to managed code.

AaronRobinsonMSFT opened the pull request.

AaronRobinsonMSFT committed:
Remove HelperMethodFrames from jithelpers

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into hmf_jithelper_throws

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
```
This list of usings looks too long.

AaronRobinsonMSFT replied:
Yep. This was copy/paste for now.

jkotas commented on src/coreclr/vm/ecall.cpp:
```diff
@@ -186,6 +186,54 @@ void ECall::PopulateManagedHelpers()
     pDest = pMD->GetMultiCallableAddrOfCode();
     SetJitHelperFunction(CORINFO_HELP_DBL2ULNG_OVF, pDest);
+
+    pMD = CoreLibBinder::GetMethod((BinderMethodID)(METHOD__THROW_HELPERS__THROWARGUMENTEXCEPTION));
```
To your earlier question about this being executed during startup for no good reason, it should be a straightforward change to do this lazily only once somebody asks for the address of the JIT helper (typical app will never ask for address for most of these throwing helpers):

- There are the 3 places in jitinterface.cpp that needs to be updated to call the lazy initialization:

            Precode* pPrecode = Precode::GetPrecodeFromEntryPoint((PCODE)hlpDynamicFuncTable[dynamicFtnNum].pfnHelper);

        pfnHelper = hlpDynamicFuncTable[dynamicFtnNum].pfnHelper;

        pfnHelper = hlpDynamicFuncTable[((size_t)pfnHelper - 1)].pfnHelper;

- The lazy initialization would convert CorInfoHelpFunc to BinderMethodID, get the helper entrypoint (`CoreLibBinder::GetMethod`, `GetMultiCallableAddrOfCode`) and cache the pointer in `hlpDynamicFuncTable`

AaronRobinsonMSFT replied:
Done. This increased the PR size a bit, but it does seem to remove some cruft and reduce any concern I had with start-up impact.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -5188,7 +5188,7 @@ BOOL MethodTable::FindDispatchEntry(UINT32 typeID,
 #ifndef DACCESS_COMPILE
 
-void ThrowExceptionForAbstractOverride(
+void ThrowEntryPointNotFoundException(
```
The contract for this method is wrong. It throws and allocates, so it cannot be LIMITED_METHOD_CONTRACT.

AaronRobinsonMSFT replied:
Yep. I've a bit of clean-up to do in this PR before it is ready.

AaronRobinsonMSFT committed:
Clean-up

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -5193,7 +5193,13 @@ void ThrowEntryPointNotFoundException(
     MethodDesc *pInterfaceMD)
 {
-    LIMITED_METHOD_CONTRACT;
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
```
```suggestion
    STANDARD_VM_CONTRACT;
```

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -5400,7 +5406,13 @@ void ThrowAmbiguousResolutionException(
     MethodDesc *pInterfaceMD)
 {
-    LIMITED_METHOD_CONTRACT;
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
```
```suggestion
    STANDARD_VM_CONTRACT;
```

AaronRobinsonMSFT committed:
Review feedback

AaronRobinsonMSFT committed:
Change JIT helper definition macro to encode BinderMethodID.
Remove CorInfoHelpSig and replace with BinderMethodID in macros.
Convert start up loading of managed JIT helpers to lazy model.

AaronRobinsonMSFT committed:
Remove the early start-up registration of managed JIT helpers.

AaronRobinsonMSFT commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5256,28 +5027,55 @@ VMHELPDEF hlpDynamicFuncTable[DYNAMIC_CORINFO_HELP_COUNT] =
+    hlpDynamicFuncTable[ftnNum].pfnHelper = (void*)pFunc;
+}
+
+VMHELPDEF LoadDynamicJitHelper(DynamicCorInfoHelpFunc ftnNum)
```
This is the main lazy load logic that was added.

AaronRobinsonMSFT commented:
This is going to fail for R2R images during `Module::RunEagerFixups()` - see below. I wonder if this can be narrowed to be "valid" if the image is SPCL?

https://github.com/dotnet/runtime/blob/e95eaf06bdc2d5faeb87b586c57d301fb7f0be43/src/coreclr/vm/ceeload.cpp#L3470-L3487

AaronRobinsonMSFT committed:
Narrowly permit type loading in SPCL during eager fixup for R2R images.

jkotas commented on src/coreclr/vm/jithelpers.cpp:
```diff
@@ -5238,16 +5001,24 @@ enum __CorInfoHelpFunc {
+};
+
+// dynamic helpers to Binder ID mapping - See definition of DynamicCorInfoHelpFunc.
+static BinderMethodID hlpDynamicToBinderMap[DYNAMIC_CORINFO_HELP_COUNT] =
```
```suggestion
static const BinderMethodID hlpDynamicToBinderMap[DYNAMIC_CORINFO_HELP_COUNT] =
```

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10550,8 +10550,8 @@ void* CEEJitInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,         /* IN  */
-        NOTHROW;
-        GC_NOTRIGGER;
+        THROWS;
+        GC_TRIGGERS;
```
This needs to be `JIT_TO_EE_TRANSITION` now (without _LEAF).

AaronRobinsonMSFT committed:
Feedback

AaronRobinsonMSFT committed:
Match casing and style of new THROWHELPER type to CASTHELPERS

AaronRobinsonMSFT committed:
Fix-up thread mode. This is a temporary fix.

jkotas reviewed (approved):
LGTM. Thanks!

AaronRobinsonMSFT committed:
Change thread mode contract to expected.

AaronRobinsonMSFT committed:
Attempt to narrow contracts

AaronRobinsonMSFT committed:
Merge branch 'main' into hmf_jithelper_throws

AaronRobinsonMSFT commented:
/ba-g All unrelated failures.
